### ISSUE TITLE
Automatic function rewriting breaks autowrap

### ISSUE THREAD

**Issue by clinssen** [sympy/sympy#23417]
We are using autowrap on an expression that contains the Heaviside step function. To help autowrap, we redefine it as ``(sympy.symbols("x") + abs(sympy.symbols("x"))) / (2 * abs(sympy.symbols("x")) + 1E-300)`` and pass it via the ``helpers`` parameter of ``autowrap.autowrap()``.

Since commit 89ffccea46, these functions are being rewritten before they have a chance to be picked up by the helper definitions. Hence, we are seeing the error 
```
ValueError: All Piecewise expressions must contain an (expr, True) statement to be used as a default condition. Without one, the generated expression may not evaluate to anything under some condition.
```
But we are exactly trying to avoid using Piecewise by using ``helpers``.

I found the commit using a git bisect on our local test, but I can produce a minimal test case that works until 89ffccea46 but fails after, if needed.

**Comment by ThePauliPrinciple** [comment]
Would you be able to provide a (minimal) example? In general, SymPy is not expected to give the same structural expressions between versions, just mathematically equivalent expressions. 

Without more details it isn't clear to me whether the `helpers` are for some reason not dealing correctly with"mathematical equivalence" on the SymPy side (in which case this should be changed).

In general, I would say that the best way to ensure that `Piecewise` is not used, is by subclassing a `Printer` and remove `Piecewise` from the known functions or alternatively, redefine `print_Piecewise` to print something which is supported for your use-case. Then have autowrap use that printer.

Maybe you can also give some details about why `Piecewise` should be avoided here.

**Comment by clinssen** [comment]
Please find a minimal reproducer below. It throws an exception on sympy >= 1.10 but works on earlier versions. I was under the impression that this was exactly what the ``helpers`` parameter was supposed to support (as an alternative to subclassing a ``Printer``). But perhaps I should do the rewriting "by hand" before passing the expression to autowrap, and not use ``helpers`` at all?

```Python
import sympy
import sympy.utilities.autowrap

foo = sympy.symbols("foo")

expr = -foo + .1 * sympy.Heaviside(-foo)
helpers = [("Heaviside", (sympy.symbols("x") + abs(sympy.symbols("x"))) / (2 * abs(sympy.symbols("x")) + 1E-300), [sympy.symbols("x")])]

sympy.utilities.autowrap.autowrap(expr,
                                    args=[foo],
                                    backend="cython",
                                    helpers=helpers)
```

**Comment by oscargus** [comment]
Not sure where this happens, but as I have introduced the automatic rewrites, I feel a bit guilty. I hope to have some time to fix it this week and it should then make it to 1.11. 

(And, yes, replacing before autowrapping should solve it, although it is indeed not ideal.)

**Comment by clinssen** [comment]
Much obliged!

**Comment by oscargus** [comment]
This turned out to be more tricky than expected... In #23666 I added two things to remedy this:
1. The ability to set a list of functions to not automatically rewrite
2. A different Piecewise for Heaviside that can be used in `autowrap`

So it should not be required to provide a helper anymore.

However, this is not really fixing your particular problem. The thing is that it is not so obvious from the code structure in autowrap how to extract the functions and set this list. Or rather, both are simple, but they are quite a few calls from each other so either one must pass this on from a rather high level to the actual time the code is generated or I must understand the code better and use some attribute of the appropriate class.

**Comment by oscarbenjamin** [comment]
How did this previously succeed?

The code here looks all wrong for the inputs you are providing:
https://github.com/sympy/sympy/blob/18afc6b4d724ed286381dcb6f0f1994578669e32/sympy/utilities/autowrap.py#L640-L647

If I try this in the debugger I get here:
```
In [1]: import sympy
   ...: import sympy.utilities.autowrap
   ...: 
   ...: foo = sympy.symbols("foo")
   ...: 
   ...: expr = -foo + .1 * sympy.Heaviside(-foo)
   ...: helpers = [("Heaviside", (sympy.symbols("x") + abs(sympy.symbols("x"))) / (2 * abs(sympy.symb
   ...: ols("x")) + 1E-300), [sympy.symbols("x")])]
   ...: 
   ...: sympy.utilities.autowrap.autowrap(expr,
   ...:                                     args=[foo],
   ...:                                     backend="cython",
   ...:                                     helpers=helpers)
> /home/oscar/current/sympy/sympy.git/sympy/utilities/autowrap.py(646)autowrap()
-> if expr.has(expr_h):
(Pdb) l
641  	    for name_h, expr_h, args_h in helpers:
642  	        helps.append(code_gen.routine(name_h, expr_h, args_h))
643  	
644  	    for name_h, expr_h, args_h in helpers:
645  	        breakpoint()
646  ->	        if expr.has(expr_h):
647  	            name_h = binary_function(name_h, expr_h, backend='dummy')
648  	            expr = expr.subs(expr_h, name_h(*args_h))
649  	    try:
650  	        routine = code_gen.routine('autofunc', expr, args)
651  	    except CodeGenArgumentListError as e:
(Pdb) p helpers
[('Heaviside', (x + Abs(x))/(2*Abs(x) + 1.0e-300), [x])]
(Pdb) p expr
-foo + 0.1*Heaviside(-foo)
(Pdb) p expr_h
(x + Abs(x))/(2*Abs(x) + 1.0e-300)
(Pdb) p expr.has(expr_h)
False
```
Presumably this should be checking if the expression has a Heaviside rather than looking for the replacement expression.

Are you sure that this is the right format for the `helpers` argument?

**Comment by oscargus** [comment]
My impression was that `helpers` is used to extract common subroutines rather than inlining them. (At least based on that piece of code.) But I realize that it wouldn't really work that well to use `has`.

The tests, e.g. https://github.com/sympy/sympy/blob/18afc6b4d724ed286381dcb6f0f1994578669e32/sympy/external/tests/test_autowrap.py#L107-L112
seems to indicate that it indeed works that way. The documentation says:
https://github.com/sympy/sympy/blob/18afc6b4d724ed286381dcb6f0f1994578669e32/sympy/utilities/autowrap.py#L576-L583
so it seems to be a current use case, although it actually should have been required to start with for `Heaviside`.

I guess that redefining existing functions in this step is a border-line case. Still, the ability to disable rewrites can be worthwhile.

**Comment by oscarbenjamin** [comment]
Apparently that test is based on a user who didn't understand how to use helpers:
https://github.com/sympy/sympy/blob/18afc6b4d724ed286381dcb6f0f1994578669e32/sympy/external/tests/test_autowrap.py#L149-L152

What does it even mean to say `helpers=('helper', a - b + c, (a, b, c))`? I don't see what effect that has. Presumably the goal is to replace `a - b + c` with something but  what should it even be replaced with.

I don't see even a trivial test that demonstrates the purpose of the `helpers` argument.

**Comment by oscarbenjamin** [comment]
It would be nice to fix this but honestly I can't figure out what the helpers argument is supposed to do. I don't think it is really supposed to be used for rewriting something like Heaviside but rather for something else except I'm not really sure what...

In any case an easier way of doing what you're doing with helpers is just this:
```python
In [8]: expr = -x + .1 * Heaviside(-x)

In [9]: w = Wild('w')

In [10]: expr
Out[10]: -x + 0.1⋅θ(-x)

In [11]: expr.replace(Heaviside(w), (w + Abs(w))/(2*Abs(w) + 1.0e-300))
Out[11]: 
      0.1⋅(-x + │x│) 
-x + ────────────────
     2⋅│x│ + 1.0e-300
```
Then you can call autowrap afterwards.

**Comment by oscarbenjamin** [comment]
Btw I notice here that it mentions a regression in sympy 1.5:
https://ode-toolbox.readthedocs.io/en/master/

Does that regression already have an open sympy issue? What specifically slowed down?

**Comment by oscargus** [comment]
> What does it even mean to say `helpers=('helper', a - b + c, (a, b, c))`?

As I understand it, it will create a subroutine called `helper` which computes `a - b + c` and extract all instances of a - b + c from the code (which is where I think it is a bit strange that it uses `has`, one would rather prefer some sort of matching so that `d - e + f` will also use that subroutine).

**Comment by oscarbenjamin** [comment]
> some sort of matching

Yes, exactly. I would have thought something like the wild replace pattern I showed above would be better.

But then why do you need the helpers argument if you can just call replace yourself?

**Comment by oscarbenjamin** [comment]
@clinssen it is not clear to me that the use of the `helpers` argument as shown above is something that is supposed to work.

I can see several possible resolutions to this issue (not mutually exclusive):

1. The way to accomplish this should just be to use replace as shown above (https://github.com/sympy/sympy/issues/23417#issuecomment-1163606795)
2. Autowrap could be changed to have a better way of handling something like this.
3. There could/should be a way to disable rewrites.

Does using `replace` work in your use case?

**Comment by oscarbenjamin** [comment]
I'm going to consider this fixed as far as the 1.11 release is concerned because we need to get the release out and as far as I can tell the OP issue is fixed on both master and the 1.11 release branch.

**Comment by clinssen** [comment]
Many thanks @oscarbenjamin, and sorry for not being able to test sooner, I'm on an extended vacation until mid-September.

**Comment by clinssen** [comment]
This seems to be fixed on the latest sympy release (1.13.1), hence this issue can be closed. Thanks for the fix!
